expo/expo
# [eyw][web] add createWebpackConfig export for expo-yarn-workspaces

# Why

The expo-yarn-workspaces example doesn't work well out of the box for web projects, mainly because the webpack configuration isn't following workspace symlinks.

Issue: https://github.com/expo/examples/issues/264

The first issue is that updates in shared packages are not reflected in web builds - you need to kill and restart the packager to see changes. This is happening because webpack is not following the symlinks found in the root `node_modules` folder of a workspace project. 

The second issue is that any javascript/typescript from a shared package is not transpiled and needs to be configured with webpack by specifying the module path in the `dangerouslyAddModulePathsToTranspile` option of the config

By exporting a function that takes care of some of this configuration for developers, getting up and running with workspaces will be a lot more user friendly for people building with expo-web.

# How

Specifying the webpack flag `config.resolve.symlinks = true` ensures that the actual files are being watched, and so changes in these files is registered.

Most projects that take advantage of EYW will likely want to use and export React components, so this PR provides this capability out of the box by adding these paths to the `dangerouslyAddModulePathsToTranspile` automatically. Note that the paths to be transpiled are the fully resolved paths, not the symlinks, so that updates are still registered properly - this bit of configuration can be a bit confusing to get working if you scope your packages under different aliases

Before: 
```
const createExpoWebpackConfigAsync = require("@expo/webpack-config");

module.exports = async function (env, argv) {
  const config = await createExpoWebpackConfigAsync({
    ...env,
    babel: {
      dangerouslyAddModulePathsToTranspile: [
        "first-package",
        "second-package",
      ]
    }
  }, argv);

  config.resolve.symlinks = true;
  
  return config;
};

```

After: 
```
const { createWebpackConfigAsync } = require("expo-yarn-workspaces/webpack");

module.exports = async function (env, argv) {
  const config = await createWebpackConfigAsync(env, argv)
  return config;
};

```

# Test Plan

`createWebpackConfigAsync`: 

- should automatically work with scoped packages (e.g `@my-project/core`)
- allow JSX exports from any shared package to work out of the box
- updates should be reflected right away when shared package source code is changed
- allows users to specify any addition node modules to be transpiled via the `env.babel.dangerouslyAddModulePathsToTranspile` option


1. Create a project from examples temlate `npx create-react-native-app --template with-yarn-workspaces`
2. Run `yarn first-app --web` project
3. Make update to first-package -> note there is no change until you kill and restart the packager
4. Export JSX from first-package -> note webpack failure
5. Add webpack.config.js in first-app createWebpackConfig: 
```
const { createWebpackConfigAsync } = require("expo-yarn-workspaces/webpack");

module.exports = async function (env, argv) {
  const config = await createWebpackConfigAsync(env, argv)
  return config;
};

```
6. Make update to first-package -> note its reflected in app 
7. Update the scope of first-package to a new alias
8. Restart and note everything still works


# Checklist
